### PR TITLE
Initial work on light op acks

### PR DESCRIPTION
**Pull request by kian-thompson**
# [AB#2228](https://dev.azure.com/fluidframework/internal/_workitems/edit/2228)

This development contains changes that will allow ODSP to send back an op ack without contents.

1. Delta connection will let ordering service know it can send back light op acks (for ODSP, this is in `supportedFeatures["api_light_op_acks"]`)
2. Delta connection will store the contents of its ops
3. Delta connection processing will fill in the contents of the op if matching clientSequenceNumber and clientId

**TODO**

### CODE REVIEW

File: packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts:256]: I'd suggest not using feature gate here, i.e. shift it to service. I think this code should just tell service that client is capable of working with such content, and service should make a call if it wants to leverage that.

Comment (kian-thompson) [reply packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts:256]: Is the best way to do that through the `socket.emit(...)` call? My thinking was that service might not know to interpret this, so I added the flag to wait until it can

File: packages/common/driver-definitions/src/storage.ts
Comment (anthony-murphy) [inline packages/common/driver-definitions/src/storage.ts:263]: i'm not totally sure why we need a new method here. Based on the discussion in the issue its not clear what the design here is. It would be good to articulate the design in the description so its available for open source.

File: packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts
Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts:256]: I think the best flow is the following:
1. Client advertise (on connection - "connect_document" / IConnect) that it supports certain feature 
2. Ordering service instructs client ("connect_document_success" / IConnected)
3. Client changes behavior based on # 2.

Here, I believe, the only client change in behavior is tracking more state in memory, i.e. there is really no protocol changes in terms of client outbound traffic. Only inbound traffic shape is changed - service does it in response to # 1.

File: packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts:667]: I'm a bit at a loss here re flow.
I believe calling into super, and emitting ops directly will result in same ops being emitted on the socket twice, right?
I'm also not sure why behavior should be different whether new feature is supported by service or not. The only difference on sending path is - should we remember extra content for future or not.

File: packages/loader/container-loader/src/connectionManager.ts
Comment (vladsud) [inline packages/loader/container-loader/src/connectionManager.ts:952]: I do not think this layer should know anything about it (assuming we are tasking driver to do all the heavy lifting - there are other options, including putting it all into this layer, or into container runtime layer).

I think you are probably thinking about ops potentially coming from multiple sources (storage, relay service).
But all ops from storage will come with content. Only ops on the same connection have a chance to not have content, as relay service can only sent "light" ops if it knows it's same client, i.e. clientId of sender and receiver are the same - i.e. the scope is single connection. Front-end (alfred) can do this kind of optimization only, and only in context of a single client connection (i.e., it may strip content only right befor sending op on the socket)

So I think all the logic can be encapsulated within DocumentDelteaConnection layer.

File: packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts:12]: It would be super-great if majority of the logic was in shared DocumentDeltaConnection class.
While we will enable it initially in ODSP, we want all services to be able to do so soon after.

**Comment by GaryWilber** [comment]
If the user has 2 clients (main client & summarizer client), would the server be able to do light op's for them? Is the client storing op content in a way that both clients can access?

**Comment by kian-thompson** [comment]
> If the user has 2 clients (main client & summarizer client), would the server be able to do light op's for them? Is the client storing op content in a way that both clients can access?

@GaryWilber I'm fairly certain the separate clients hold separate connections, so they would be tracking op contents individually.

File: packages/drivers/driver-base/src/documentDeltaConnection.ts
Comment (vladsud) [inline packages/drivers/driver-base/src/documentDeltaConnection.ts:100]: content should always be a string. Typing on messages sucks, but if we can make it string everywhere else, that would be  nice. That said, in the future we will support array buffer as well, so maybe keeping it as-is is fine

File: packages/drivers/driver-base/src/documentDeltaConnection.ts
Comment (vladsud) [inline packages/drivers/driver-base/src/documentDeltaConnection.ts:166]: Not sure if code becomes easier to read, but you can move this check up, i.e.
if (event === "op" && this.lightOpAcksEnabled) {
   // more complicated path
} else {
    listener = (...args: any[]) => { this.emit(event, ...args); };
}

File: packages/drivers/driver-base/src/documentDeltaConnection.ts
Comment (vladsud) [inline packages/drivers/driver-base/src/documentDeltaConnection.ts:171]: Not sure if we will need it, but you can make it a bit more prepared for potential future optimization (not sure even if they are possible, but it does not make things worse, I think) and check not clientId, but docId.
That way, maybe, we will be able to make it work for case of interactive client + summarizer client on same socket in the future.

File: packages/drivers/driver-base/src/documentDeltaConnection.ts
Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:172]: Also, if there is such opportunity - it would be nice to assert that we are not leaking memory by accumulating (and not removing) entries form this.opContent. One place where it actually happens (and possibly pointing to us changing design) is noops and coalescing by service - client sends noops, but service may never sequence them (and thus skip localSequenceNumber). Over time we might accumulate a lot of them.

That all said, ops on the wire can come out of order (I think, at least we do not make any assumptions about order anywhere in the stack). If that's the case, cleaning this.opContent becomes very hard (as there no definitive signal or event that we can use to do so) Maybe we disregard my suggestion on being defensive here (prev comment) and simply not remember ops with empty content? And at the time of disconnect simply verify that we have no more than X entries in this.opContent? (X would need to be vey large, unfortunately, so usefulness of this check would be very low).

Let's brainstorm it together, I have no good suggestion here other than raising the problem

File: packages/drivers/driver-base/src/documentDeltaConnection.ts
Comment (kian-thompson) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts:100]: The `contents` type is currently `any` for both `IDocumentMessage` and `ISequencedDocumentMessage`

File: packages/drivers/driver-base/src/documentDeltaConnection.ts
Comment (vladsud) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts]: I believe docId here is PUSH docId, and it uniquely identifies document for this connection. Can you please poke around? I believe there are other places where we do comparisons with it, so hopefully it becomes clearer on what it tracks - you can always ask Gary, but probably we want to follow up with AFR folks as well.

Comment (kian-thompson) [reply packages/drivers/driver-base/src/documentDeltaConnection.ts]: In the example I was running the socket/document ID was the same across different client IDs, so I think we have to stick with clientId for now

**Review by vladsud — APPROVED** [review]
LGTM. But please follow up with Gary and have follow up items to test it once we have service implementation. From roll out, it's possible it's easier if we actually control enabling from client side, i.e. check in this code (or use final feature names) only after PUSH shipped its implementation. This way, we can enable it first in our stress tests before it hits production (vs. if we start with client, Gary's enabling feature on his side results in immediate hit to some internal users)

**Comment by kian-thompson** [comment]
[AB#2959](https://dev.azure.com/fluidframework/internal/_workitems/edit/2959) tracks follow-up that will merge this PR and do the testing

**Comment by vladsud** [comment]
@kian-thompson, anything blocks this PR?

**Comment by vladsud** [comment]
@kian-thompson - anything blocking this PR?

**Comment by kian-thompson** [comment]
> @kian-thompson - anything blocking this PR?

@vladsud 
Sorry, didn't see the other question a couple weeks ago. I was going to wait for service side to finish their implementation before merging this code into main [AB#2959](https://dev.azure.com/fluidframework/internal/_workitems/edit/2959)

**Comment by kian-thompson** [comment]
Closing this PR as it's been stale for a while

**Pull request closed without merging**
